### PR TITLE
improvements to the HGCHEback digitizer and implementation of a end-of-life scenario for the HGC scintillator section

### PR DESCRIPTION
**Pull request by deguio**
this is a copy of PR #26380 which I cannot re-open.
reporting here PR description. sorry for the troubles..

------------------------------
#### PR description:

this PR implements a more realistic version of the HGC HE-back digitizer:
- updated parameters, closer to the current design of the HGC HE-back section, are included
- switches are added to choose the digi algorithm and enable the dependence of the signal on the scintillator tile geometry
- a first implementation of a end-of-life scenario (3000/fb, 16.7kh) for HGC HE-back is implemented. the scenario is automatically enabled when the appropriate custom function defined in `aging.py` is called
- a standalone analyzer is added to allow a simplified plotting of the dose, fluence and SoN maps

the references that support this implementation are:
- [Notes from Philippe (mid 2018)](https://deguio.web.cern.ch/deguio/hgcal/docs/Note_area.pdf)
- [DN from Jeremy/Pawel (early 2017)](cms.cern.ch/iCMS/jsp/openfile.jsp?tp=draft&files=DN2017_010_v1.pdf)
- [Notes from Chris (early 2019)](https://deguio.web.cern.ch/deguio/hgcal/docs/ElectronicsNotes.pdf)
- [Dose and Fluence parametrization by Jeremy (April 2019)](https://deguio.web.cern.ch/deguio/hgcal/docs/scintillator-fits.pdf)
- CMS-TDR-019

This set of changes will allow to study the evolution of occupancy and bandwidth for a end-of-life configuration of the HGC scintillating section.

#### PR validation:

A full validation was presented at the [HGCal DPG](https://indico.cern.ch/event/810242/#2-simulation-of-scintsipms)

in summary:
- small differences are expected for a non-aged scenario
- an improved description of the signal and noise evolution is included depending from cell location, scintillator darkening and dark current for a end-of-life scenario
- the implementation is not expected to affect dramatically the recHit energy scale

in terms of performance here is the impact of this PR on the HGC HE-back digitization:
- [current version](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo1.pp/105)
- [new implementation, begin of life](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo2.pp/116)
- [new implementation, end of life](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo2_3000_geom.pp/31)
- [new implementation, end of life, after commit 4a52c5eb25d](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo2_3000_geom.4a52c5eb25d.pp/33)
- [new implementation, end of life, after commit dc5acc22b2b](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo2_3000_geom.dc5acc22b2b.pp/55)
- [new implementation, v10 on 10_6_0_pre4, end of life, after commit 5c42c3cad85](http://deguio.web.cern.ch/deguio/cgi-bin/igprof-navigator/HGCalDigi/outputAlgo2_3000_geom.5c42c3cad85.pp/63)

@pfs @perrotta

### CODE REVIEW

**Comment by perrotta** [comment]
Before launching the tests, please @deguio apply the fixes in the PSet in
https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1THGCal/python/hgcalVFEProducer_cfi.py#L50
(as it was modified after #26545): in this PR the paramenter HGCAL_noise_MIP is renamed

Please also run some test and verify whether there are also other modification from #26545 or from some other recent commit that need to be fixed to comply with what you are implementing here,

**Comment by deguio** [comment]
I've pushed the change and tested WF 29034.0 and it runs fine.
however from what I see the [vfe_proc](https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1THGCal/python/hgcalVFEProducer_cfi.py#L37) as it is doesn't see the effect of the functions in `hgcalDigitizer_cfi.py` in case they are called.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> however from what I see the [vfe_proc](https://github.com/cms-sw/cmssw/blob/master/L1Trigger/L1THGCal/python/hgcalVFEProducer_cfi.py#L37) as it is doesn't see the effect of the functions in `hgcalDigitizer_cfi.py` in case they are called.

What do you mean? That if you switch to, e.g., ```HGCal_setEndOfLifeNoise``` or ```HGCal_disableNoise``` this will not get propagated to hgcalVFEProducer_cfi.py? Please explain. 
If so, this has to be fixed for L1: maybe @jbsauvan can help with it and provide a commit

**Comment by deguio** [comment]
@perrotta yes, that's what I meant.

**Comment by jbsauvan** [comment]
I can check again, but I think it was correctly propagated.

**Comment by perrotta** [comment]
> @perrotta yes, that's what I meant (it has nothing to do with this PR).

Thank you @deguio 
I agree that it has nothing to do with this PR. But, since it was noticed here, I would prefer to have it fixed now instead of risking to forget it. You now need a L1 signature anyhow here, and if the fix is simple it can even get quickly included here without delaying the approval. Otherwise, it can go in a separate PR.

**Comment by jbsauvan** [comment]
Actually Era modifications are correctly propagated, but not the changes from aging customisation.
I suppose in that case we have to pass parameters as `cms.PSet(refToPSet_ = ...)` to propagate things correctly, but is there a better syntax available to do that?

**Comment by jbsauvan** [comment]
@perrotta I've implemented a fix to propagate correctly noise value changes to the VFE producer. I'll make a PR on @deguio branch.

**Comment by deguio** [comment]
thanks @jbsauvan

**Comment by perrotta** [comment]
please test

**Comment by deguio** [comment]
the data workflows are failing because of:
`AttributeError: 'Process' object has no attribute 'HGCAL_noise_fC'`

@jbsauvan @perrotta please suggest how you prefer to proceed. should I revert
`https://github.com/deguio/cmssw/pull/2`?

**Comment by jbsauvan** [comment]
These errors are only in data workflows; it seems that HGCAL trigger primitives configs are loaded from `Configuration.StandardSequences.SimL1EmulatorRepack...` importing `L1Trigger.Configuration.SimL1Emulator`, which imports HGCAL TPG configs and applies them to HGCAL eras. 
But I don't know what would be the best way to prevent loading HGCAL TPG stuff for non HGCAL eras.

**Comment by jbsauvan** [comment]
Since it is linked to the trigger configuration at a higher level than HGCAL, maybe @rekovic can suggest a way to fix this.

**Comment by perrotta** [comment]
> @jbsauvan @perrotta please suggest how you prefer to proceed. should I revert
> `https://github.com/deguio/cmssw/pull/2`?

The bug in the L1 configuration must definitely get fixed: there is no point that L1 cannot comply to the different scenarios foreseen for the HGCal.

On the other hand, that issue is not coming from the original formulation of this PR: therefore, while I would prefer that a fix can be found quickly enough and implemented already here, if this is not the case I agree that the last commit can be reverted and the issue left in the L1/upgrade hands. 

Let say: if there is no fix available by Monday evening let revert it.

@kpedro88 @rekovic

**Comment by makortel** [comment]
> But I don't know what would be the best way to prevent loading HGCAL TPG stuff for non HGCAL eras.

The framework group suggests to not go for conditional loading/importing. A possible solution would be to move the PSet definitions out from `SimCalorimetry/HGCalSimProducers/python/hgcalDigitizer_cfi.py` to a place that can be loaded also by data L1REPACK configurations. @kpedro88 may have some further suggestions (we talked about this).

I could also note that the exception messages from the failed jobs lead to improving them in #26850.

**Comment by kpedro88** [comment]
I'll try to take a look at this tomorrow to see if I can find an immediately-implementable solution.

@makortel reminded me that the "right" way to have shared parameters is an ESProducer (if a database object is not available, which is the case for HGCal currently). It would be good to migrate these HGCal parameters to one or more ESProducers at some point. We previously did this for HCAL aging parameters.

**Comment by deguio** [comment]
hi @kpedro88 
that's on our list. it will come in a future pr.

**Comment by kpedro88** [comment]
@deguio I sent you https://github.com/deguio/cmssw/pull/3, which is the simplest way I can find to solve this issue right now.

**Comment by kpedro88** [comment]
@deguio I'm also concerned that the parent pset is cloned when it's sent to the producer:
https://github.com/cms-sw/cmssw/blob/0ea25e1a9b3df66cfd4ed5b6eee1778113406a9a/L1Trigger/L1THGCal/python/hgcalVFEProducer_cfi.py#L68

Can you double-check after taking my PR that the aging customizations are propagated appropriately?

**Comment by jbsauvan** [comment]
Thanks @kpedro88 I'll cross-check Tomorrow morning.

**Comment by makortel** [comment]
@kpedro88 Did you test it (on a data workflow)? By just browsing the code I'd expect the module construction to throw an exception (e.g.) in
https://github.com/cms-sw/cmssw/blob/0ea25e1a9b3df66cfd4ed5b6eee1778113406a9a/L1Trigger/L1THGCal/src/veryfrontend/HGCalVFESummationImpl.cc#L13-#L14

**Comment by kpedro88** [comment]
so far I only tested to check that the python error went away, I'll run the WF now (the module shouldn't actually run in a non-phase 2 workflow...)

**Comment by makortel** [comment]
> (the module shouldn't actually run in a non-phase 2 workflow...)

True, but if the module is defined in the configuration, it will be constructed (and the L1REPACK+HLT steps do not run `convertToUnscheduled()` so modules not in any Path/EndPath (also via Sequences/Tasks) do not get pruned.

**Comment by kpedro88** [comment]
136.731 step 2 ran without a problem.

**Comment by makortel** [comment]
> 136.731 step 2 ran without a problem.

Thanks. I also verified independently that my assumption

>  if the module is defined in the configuration, it will be constructed 

was incorrect (i.e. modules not attached in Paths/EndPaths are not constructed).

**Comment by jbsauvan** [comment]
I checked that the noise customizations are correctly propagated with @kpedro88 update

**Comment by deguio** [comment]
thanks @kpedro88

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Changes in reco code are minor and consequential to the changes in simulation
- Jenkins outputs and validation plots provided reflect all that
- Performance as reported in the PR description in terms of CPU is good

**Comment by kpedro88** [comment]
+upgrade

The noise distribution in the scintillator part of HGCal changes as expected:
HEB energy

The energy distributions in the silicon parts of HGCal (EM and hadronic) just have slight fluctuations from random number changes in the DIGI step:
EE energy
HEF energy

File: SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc [hunk 1]
```diff
@@ -11,27 +11,312 @@
...
+                                      << scaleFactor << std::endl;
+  }
+
+  return scaleFactor;
+}
+
+std::array<double, 8> HGCHEbackSignalScaler::computeRadius(const HGCScintillatorDetId& cellId)
+{
+  GlobalPoint global = hgcalGeom_->getPosition(cellId);
+
```
Comment (civanch) [inline SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: @deguio , sorry for the late comment but it seems std::pow is not needed in this method, it is only overhead.

Comment (deguio) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: hi @civanch 
are you suggesting to use plain multiplications instead of `std::pow`?

Comment (civanch) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: @deguio , yes, exactly

Comment (kpedro88) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: Modern compilers optimize `pow(x,2)` to be the same as `x*x`, but not sure about other exponents

Comment (makortel) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: Quick tests with godbolt indicate that GCC 7/8/9 and clang 7/8 compile `pow(x, 3)` to an actuall call to `pow()`.

Comment (deguio) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: ok. so what's the suggestion? change to plain multiplication for the third degree exponents?

Comment (kpedro88) [reply SimCalorimetry/HGCalSimProducers/src/HGCHEbackDigitizer.cc]: Best would be e.g. `double radius_m100_3 = radius_m100_2*radius_m100;`

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- Changes since latest reco signature in https://github.com/cms-sw/cmssw/pull/26811#issuecomment-494812675 are minor

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade
see https://github.com/cms-sw/cmssw/pull/26811#issuecomment-494940666 for summary of changes

**Comment by fabiocos** [comment]
@rekovic could you please scrutinize this PR?

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
@deguio 
 > * small differences are expected for a non-aged scenario

Do you have more information on this, in the slides of the above presentations you listed ? Thanks

**Comment by perrotta** [comment]
@deguio : and now this has to be rebased, conflicts with RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py ...
Please @fabiocos take into account that it is fully signed now, before the needed rebase

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@deguio please confirm that the only last changes are in the RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py which had conflicts (so that we can avoid reviewing the whole PR once again...)

**Comment by deguio** [comment]
hi @rekovic 
the main differences for a non-aged detector configuration are coming from:
- updated value for `nPEperMIP` in `hgcalDigitizer_cfi.py `
- scaling of the signal depending on the tile size

you can have a look at slides 15 and 16 from the presentation to see the effect.
however, since the signal is then re-calibrated, the only visible effect is a smearing.
you can look at the plots posted by kevin [here](https://github.com/cms-sw/cmssw/pull/26811#issuecomment-494940666)

**Comment by deguio** [comment]
> @deguio please confirm that the only last changes are in the RecoLocalCalo/HGCalRecProducers/python/HGCalRecHit_cfi.py which had conflicts (so that we can avoid reviewing the whole PR once again...)

yes, I confirm

**Comment by deguio** [comment]
I'm confused about what the problem is..
any advice?
thanks.

**Comment by perrotta** [comment]
please test
(let retry with a possibly cleaner IB)

**Comment by perrotta** [comment]
+1
- Only rebased since previous reco signature in https://github.com/cms-sw/cmssw/pull/26811#issuecomment-494812675:
  - Changes in reco code are minor and consequential to the changes in simulation
  - Jenkins outputs and validation plots provided reflect all that
  - Performance as reported in the PR description in terms of CPU is good

**Comment by fabiocos** [comment]
@rekovic @civanch @kpedro88 I understand that this PR was already signed by all of you, and the latest updated after rebase is in the RECO area and is ok. I will move forward with the integration, please comment in case for further updates.

**Comment by fabiocos** [comment]
I see that this PR touches the same files as other open PRs:
```
File  RecoLocalCalo/HGCalRecProducers/interface/HGCalCLUEAlgo.h  modified in PRs #  ['27128', '26811']
File  RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h  modified in PRs #  ['27049', '27128', '26811']
File  RecoLocalCalo/HGCalRecProducers/plugins/HGCalRecHitWorkerSimple.cc  modified in PRs #  ['27129', '26811']
```
but for what I can see the lines are different, so I do not expect conflicts when this is merged

**Comment by kpedro88** [comment]
+upgrade

**Comment by civanch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**